doc/asset-contracts-cbor.md

```diff
@@ -0,0 +1,189 @@
+
+## Solution
+
+Our overall approach to solving the problems above:
+* Switch contracts from JSON to (a subset of) CBOR, a compact binary object representation which is a superset
+  of JSON (defined in RFC 7049).
+* For ease of distribution, define a standard for on-chain distribution of contracts in `OP_RETURN` outputs.
+* Everything about this standard is an optional layer on top of the Elements chain; no consensus or
+  standardness rule of Elements will require its use.
+
+* Elements Core RPC changes:
+  * Retrieve and store asset contracts from `OP_RETURN` outputs in issuance transactions.
+    * Validate against the contract hash in the issuance, and against the contract well-formedness
+      rules (which are defined below.)
+    * Maintain a whitelist of known assets with no on-chain contract, including Bitcoin.
+    * Provide new RPCs for working with (providing, retrieving) asset contracts.
+  * Significant changes to the handling of asset values in RPC calls, detailed in `asset-values-rpc.md`.
+
+* Elements GUI changes:
+  * Display, and accept user entry of, decimal values of known and whitelisted assets (using their contract-specified or whitelist-specified precision.)
+  * Do not display, or accept user entry of, values for other assets.
+  * Treat reissuance tokens specially, as always having precision 0 (that is, atomic indivisible units.)
+
+* Registry changes:
+  * Accept registraton of CBOR contracts.
+  * Provide for verification of CBOR contracts against the spec.
+  * Stop accepting new JSON contracts (perhaps with a "deprecation phase" where they are discouraged somehow).
+  * For convenience, allow retrieval of contracts either in CBOR format, or converted to JSON as described later in this document.
+
+* Tooling:
+  * `rust-elements`: support creating, inspecting, and verifying both old and new asset contracts
+  * `hal-elements`: support creating, inspecting, and verifying both old and new asset contracts
+  * TBD: Usable tooling for registering assets with the registry, and issuing assets on the blockchain.
+  * For users needing generic tooling for working with CBOR data: https://github.com/intel/tinycbor, http://cbor.me/
+
+## Contract format
+
+A version 1 asset contract consists of: a single-byte version field (which must be 1), followed by a CBOR-encoded array with 3 elements:
+* The precision, encoded as an unsigned integer from 0 to 8.
+* The ticker, encoded as a string, 3-5 characters from the character set "[a-zA-Z.-]" (that is, upper and lowercase letters, dot, and dash.)
+* A map containing other fields, all of which are OPTIONAL:
+  * (The field names `"precision"`, `"ticker"`, and `"entity"` are reserved for backwards compatibility, and may not be used.)
+  * `"name"`, a string containing ASCII characters, length 1 to 255 (inclusive)
+  * `"issuer_pubkey"`, a byte-string containing a secp256k1 public key in compressed form.
+  * `"domain"`, a domain name belonging to the issuer
```

Comment:
Maybe also specify how the accompanying real-world contract can be found on this domain, with the hash of the text of this real-world contract to be used as 'contract hash' in the issuance ?

Reply:
This way wallets will be able to locate the real-world contract governing the off-chain conditions, verify that it is indeed referenced in the issuance, and display the contract to the user

Reply:
We mention the use of the 'domain' field a bit later under 'Registry rules' -- it's only used to prove the identity of the asset issuer. The use of "contract" here is not referring to a legal contract, but just to this structure with the parameters of the asset in it. (There is some discussion below of how these "contracts" are distributed.)

If the issuer wanted to add a reference to some other document, they would do that by adding another field, but it would be up to them to define the meaning of it.

Reply:
Why not define the standard way to do this, especially if there _is_ a need to pull some file from that domain to prove the identity of the issuer ?

Reply:
and if standard way is defined, then there's no need for adding another field

Reply:
Added the details for the way the Liquid Asset Registry does this right now.